### ISSUE TITLE
ResultCode: -120

### ISSUE THREAD

**Issue by austinmroczek** [craigjmidwinter/total-connect-client#66]
ResultCode: -120. ResultData: Feature Not Supported

From https://github.com/home-assistant/home-assistant/issues/32358

a) Need to figure out why we're getting this result code.

b) Total connect client needs to handle this result code.

**Comment by austinmroczek** [comment]
@codypet 

Our total-connect-client code is not handling result code -120, which is preventing proper operation of the component in Home Assistant.  We don't have good documentation for error messages, and haven't seen this one yet. I'll work to handle that error code, however, I'm not sure why you would get that it. 

Have you enabled "sensor events" as described at https://www.home-assistant.io/integrations/totalconnect/ ?

**Comment by codypet** [comment]
@austinmroczek yes.  They're all toggled on.

**Comment by austinmroczek** [comment]
@codypet  I'd like to see info about your system if possible.  I want to see if there's something in your system/location/user "flags" that would explain why you're getting -120 when others are not.

Download these two files (which I just modified to show more relevant info):
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/TotalConnectClient.py`
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/test.py`

Then run test.py with:
`python3 test.py username password -1`

Make sure to replace sensitive info with "x" or some other character.

**Comment by codypet** [comment]
Running test.py yields the response below.  Line 6, what is Zeep?
![image](https://user-images.githubusercontent.com/17790157/75629070-ef914900-5bac-11ea-873a-051f4b3bbd75.png)

**Comment by austinmroczek** [comment]
`zeep` is the python package we use to handle the SOAP messages between our code and Honeywell.

Sorry I forgot to say that before running test.py you should activate your virtual environment by running `source`.  Assuming your venv setup followed the usual instructions, it should be
`source /srv/homeassistant/bin/activate`

Your virtual environment should already have `zeep` installed.

**Comment by codypet** [comment]
My mistake.  The result from test.py was this.  Key error 9?
![Screenshot_20200301-171537_Termius](https://user-images.githubusercontent.com/17790157/75635065-241ef800-5be1-11ea-8149-1fcd4ad233ba.jpg)

**Comment by austinmroczek** [comment]
Hmm...that's an odd error (or I'm bad at helping people troubleshoot).  Let's do two things:

1) Did you get a test.log file ?  It should be in the same directory.  Please post it here.  Be careful to remove user/password info. `grep -n username test.log` is your friend.

2) I already made some fixes today.  Re-download the two files and try again.

**Comment by austinmroczek** [comment]
Well I just figured out part of the problem.  You must have downloaded the code after I updated it today.  Where you previously got error -120 and the program stopped, it now logs an error and keeps going.  The KeyError 9 above is because it was expecting zone info to be filled in already...but it wasn't.  I'll fix that shortly and post an update to the code.

**Comment by codypet** [comment]
Ah good to know.  I redownloaded the files just now before I saw you seconded message and got the image below.  Running test.info gave me this
```
6210:<soap-env:Envelope xmlns:soap-env="http://schemas.xmlsoap.org/soap/envelope/"><soap-env:Body><ns0:LoginAndGetSessionDetails xmlns:ns0="https://services.alarmnet.com/TC2/"><ns0:userName>xxxxxxxxx</ns0:userName><ns0:password>xxxxxxxx</ns0:password><ns0:ApplicationID>14588</ns0:ApplicationID><ns0:ApplicationVersion>1.0.34</ns0:ApplicationVersion></ns0:LoginAndGetSessionDetails></soap-env:Body></soap-env:Envelope>
</ns0:userName><ns0:password>xxxxxxxxxxx</ns0:password><ns0:ApplicationID>14588</ns0:ApplicationID><ns0:ApplicationVersion>1.0.34</ns0:ApplicationVersion></ns0:LoginAndGetSessionDetails></soap-env:Body></soap-env:Envelope>
```
![Screenshot_20200301-200703_Termius](https://user-images.githubusercontent.com/17790157/75638291-137a7c00-5bf9-11ea-84c5-a227299608d0.jpg)

**Comment by austinmroczek** [comment]
Please post the log line (really long) that includes `LoginAndGetSessionDetailsResult` 

That should show various values about your account, and when compared to mine may show a reason why you're getting -120 and the rest of us aren't.

Here's mine:

```
<?xml version="1.0" encoding="utf-8"?><soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema"><soap:Body><LoginAndGetSessionDetailsResponse xmlns="https://services.alarmnet.com/TC2/"><LoginAndGetSessionDetailsResult><ResultCode>0</ResultCode><ResultData>Success</ResultData><ModuleFlags>Security=1,Video=1,Automation=1,GPS=1,VideoPIR=1,ReadState=1,IsAlexaEnabled=0</ModuleFlags><SessionID>xxxxxxxxxxxxxxxxxx</SessionID><UserInfo><UserID>xxxxx</UserID><Username>xxxxx</Username><Fullname>Home Assistant</Fullname><Language>0</Language><LocaleID>0</LocaleID><UserFeatureList>Master=0,User Administration=0,Configuration Administration=0</UserFeatureList><ClientPreferences /><IsEulaAccepted>true</IsEulaAccepted><IsSMSEulaAccepted>false</IsSMSEulaAccepted><DateFormatID>0</DateFormatID><TimeFormatID>0</TimeFormatID><PushNotificationStatus>0</PushNotificationStatus><HasResetPassword>true</HasResetPassword><IsRootedDeviceAccepted>false</IsRootedDeviceAccepted></UserInfo><Locations><LocationInfoBasic><LocationID>xxxxxx</LocationID><LocationName>Home1</LocationName><PhotoURL /><LocationModuleFlags>Security=1,Video=0,Automation=0,GPS=0,VideoPIR=0,TimeTriggeredEvent=0,TemperatureUnits=F,ConfigureContent=0,SyncLocation=0,ConfigureSlideshow=0,TimezoneOffset=-8.0,SmartAction=0,CustomArm=0,NoTriggerScene=0,NoScene=0,AutoSyncEnabled=1,WiFIThermostatEnabled=1,SupportsOnlyHDPhotos=0,SyncStatusSupported=1,WiFiHBSupported=1,DoorBellSupported=1,masterUserCodeSync=0,GeofenceStatus=0,RSISupported=0,VideoServiceEnabled=0,HasAddressUpdated=1,WifiGaragedoorSupported=0,OutboundServiceEnabled=0,HasSmartScenes=0,MotionViewerServiceEnabled=0,VavEnabled=0</LocationModuleFlags><SecurityDeviceID>661042</SecurityDeviceID><DeviceList><DeviceInfoBasic><DeviceID>xxxxxx</DeviceID><DeviceName>ILP5</DeviceName><DeviceClassID>1</DeviceClassID><DeviceSerialNumber>xxxxxxxx</DeviceSerialNumber><DeviceFlags>PromptForUserCode=0,PromptForInstallerCode=0,PromptForImportSecuritySettings=0,AllowUserSlotEditing=0,CalCapable=1,CanBeSentToPanel=0,CanArmNightStay=0,CanSupportMultiPartition=0,PartitionCount=0,MaxPartitionCount=0,OnBoardingSupport=0,PartitionAdded=0,DuplicateUserSyncStatus=0,PanelType=8,PanelVariant=1,BLEDisarmCapable=0,ArmHomeSupported=0,DuplicateUserCodeCheck=0,CanSupportRapid=0,IsKeypadSupported=1,WifiEnrollmentSupported=0,IsConnectedPanel=0,ArmNightInSceneSupported=0,BuiltInCameraSettingsSupported=0</DeviceFlags></DeviceInfoBasic><DeviceInfoBasic><DeviceID>xxxxxx</DeviceID><DeviceName>WiFi DoorBell</DeviceName><DeviceClassID>7</DeviceClassID><DeviceSerialNumber>xxxxxxx</DeviceSerialNumber><DeviceFlags /></DeviceInfoBasic></DeviceList></LocationInfoBasic></Locations><InterfaceSchemaConfigInfo><AppleDownloadURL /><AndroidDownloadURL /><LatestSupportedAppVersion /></InterfaceSchemaConfigInfo></LoginAndGetSessionDetailsResult></LoginAndGetSessionDetailsResponse></soap:Body></soap:Envelope>

```

**Comment by codypet** [comment]
Hmm I don't get one that says result.

**Comment by austinmroczek** [comment]
Weird.  I updated the TotalConnectClient.py so it should prevent the KeyError now.  Please try one more time with test.py

**Comment by codypet** [comment]
Still getting keyerror 2 like I was 3 hrs ago
![Screenshot_20200301-231249_Termius](https://user-images.githubusercontent.com/17790157/75644998-5b59cd00-5c12-11ea-8199-34d401258859.jpg)

**Comment by austinmroczek** [comment]
Please download the two files, which I've updated
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/TotalConnectClient.py`
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/test.py`

Then run test.py again.

**Comment by codypet** [comment]
![Screenshot_20200302-180151_Termius](https://user-images.githubusercontent.com/17790157/75726280-0bc9de80-5cb0-11ea-8f8f-3c7c0f879385.jpg)
Key error 9 again.  Weird.

**Comment by austinmroczek** [comment]
Line 523 is  
`zones = data.get("Zones")`
so you're not running the latest code for some reason.

**Comment by codypet** [comment]
How do I fix that?

**Comment by austinmroczek** [comment]
do
`rm TotalConnectClient.py test.py`
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/TotalConnectClient.py`
`wget https://raw.githubusercontent.com/austinmroczek/total-connect-client/master/total_connect_client/test.py`
`source /srv/homeassistant/bin/activate`
`python3 test.py username password -1`

**Comment by codypet** [comment]
So test.py now gives me this:
```
### UserInfo ###

OrderedDict([('UserID', 1457650), ('Username', 'XXXXXXX'), ('Fullname', 'XXXXXXX'), ('Language', 0), ('LocaleID', 0), ('UserFeatureList', 'Master=1,User Administration=1,Configuration Administration=1'), ('ClientPreferences', 'zWaveModuleControlMode%3D0%26zWaveInactivityCounter%3D1%26version%3D2%26userChoseDefaultLocation%3Dfalse%26defaultLocationID%3D%252D1%26moduleOrder%3D2%253A1%252C0%253A1%252C4%253A0%252C3%253A1%252C1%253A0'), ('IsEulaAccepted', True), ('IsSMSEulaAccepted', False), ('DateFormatID', 1), ('TimeFormatID', 0), ('PushNotificationStatus', 2), ('HasResetPassword', True), ('IsRootedDeviceAccepted', False)])

### ModuleFlags ###

Security=1,Video=1,Automation=1,GPS=1,VideoPIR=1,ReadState=1,IsAlexaEnabled=1

### Locations ###

LocationID: 752636
LocationName: XXXXXXXXX
PhotoURL: None
LocationModuleFlags: Security=1,Video=0,Automation=0,GPS=0,VideoPIR=0,TimeTriggeredEvent=0,TemperatureUnits=F,ConfigureContent=0,SyncLocation=0,ConfigureSlideshow=0,TimezoneOffset=-5.0,SmartAction=0,CustomArm=0,NoTriggerScene=0,NoScene=0,AutoSyncEnabled=0,WiFIThermostatEnabled=0,SupportsOnlyHDPhotos=0,SyncStatusSupported=0,WiFiHBSupported=0,DoorBellSupported=0,masterUserCodeSync=0,GeofenceStatus=2,RSISupported=0,VideoServiceEnabled=0,HasAddressUpdated=1,WifiGaragedoorSupported=0,OutboundServiceEnabled=0,HasSmartScenes=0,MotionViewerServiceEnabled=0,VavEnabled=0
SecurityDeviceID: 980952
DeviceList: OrderedDict([('DeviceInfoBasic', [OrderedDict([('DeviceID', 980952), ('DeviceName', 'Security Panel'), ('DeviceClassID', 1), ('DeviceSerialNumber', '00D02D79C1BB'), ('DeviceFlags', 'PromptForUserCode=0,PromptForInstallerCode=1,PromptForImportSecuritySettings=0,AllowUserSlotEditing=1,CalCapable=0,CanBeSentToPanel=0,CanArmNightStay=0,CanSupportMultiPartition=0,PartitionCount=0,MaxPartitionCount=0,OnBoardingSupport=0,PartitionAdded=0,DuplicateUserSyncStatus=0,PanelType=5,PanelVariant=1,BLEDisarmCapable=0,ArmHomeSupported=0,DuplicateUserCodeCheck=0,CanSupportRapid=0,IsKeypadSupported=1,WifiEnrollmentSupported=0,IsConnectedPanel=0,ArmNightInSceneSupported=0,BuiltInCameraSettingsSupported=0'), ('SecurityPanelTypeID', None), ('DeviceSerialText', None)])])])
AcLoss: False
LowBattery: False
IsCoverTampered: False
Arming State: 10203
```
I rebooted and still don't get TotalConnect to show.   The error log appears to be the same.
`Error during setup of component totalconnect
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/setup.py", line 174, in _async_setup_component
    component.setup, hass, processed_config  # type: ignore
  File "/usr/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/components/totalconnect/__init__.py", line 37, in setup
    client = TotalConnectClient.TotalConnectClient(username, password)
  File "/srv/homeassistant/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 81, in __init__
    self.authenticate()
  File "/srv/homeassistant/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 119, in authenticate
    self.populate_details(response)
  File "/srv/homeassistant/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 143, in populate_details
    self.get_zone_details(location["LocationID"])
  File "/srv/homeassistant/lib/python3.7/site-packages/total_connect_client/TotalConnectClient.py", line 411, in get_zone_details
    + str(result["ResultData"])
Exception: Could not retrieve zone detail data. ResultCode: -120. ResultData: Feature Not Supported`

**Comment by austinmroczek** [comment]
Since you could run test.py that means the update code I wrote now handles your -120 error.  I have pushed a new module version 0.53 to pypi.  I also requested that Home Assistant increase the version it requires to 0.53, but that change will likely not be available until 0.107.  Assuming this change is made in 0.107, your alarm panel should work, but the zone statuses would not.

I still need to digest all of the results you provided above and try to see why you get -120 in the first place.  Once we figure that out, we can get your zone statuses working.

**Comment by austinmroczek** [comment]
Here's what is different about our results:

@codypet result
### UserInfo ###
OrderedDict([('UserFeatureList', 'Master=1,User Administration=1,Configuration Administration=1'), ('ClientPreferences', 'zWaveModuleControlMode%3D0%26zWaveInactivityCounter%3D1%26version%3D2%26userChoseDefaultLocation%3Dfalse%26defaultLocationID%3D%252D1%26moduleOrder%3D2%253A1%252C0%253A1%252C4%253A0%252C3%253A1%252C1%253A0'), ('DateFormatID', 1), ('PushNotificationStatus', 2)])

### ModuleFlags ###
IsAlexaEnabled=1

### Locations ###
LocationModuleFlags: AutoSyncEnabled=0,WiFIThermostatEnabled=0,SyncStatusSupported=0,WiFiHBSupported=0,DoorBellSupported=0,masterUserCodeSync=0,GeofenceStatus=2
DeviceList: OrderedDict([('DeviceInfoBasic', [OrderedDict([('DeviceFlags', 'PromptForInstallerCode=1,AllowUserSlotEditing=1,CalCapable=0,PanelType=5')])])])


@austinmroczek result
### UserInfo ###
OrderedDict([('UserFeatureList', 'Master=0,User Administration=0,Configuration Administration=0'), ('ClientPreferences', None), ('DateFormatID', 0), ('PushNotificationStatus', 0)])

### ModuleFlags ###
IsAlexaEnabled=0

### Locations ###
LocationModuleFlags: AutoSyncEnabled=1,WiFIThermostatEnabled=1,SyncStatusSupported=1,WiFiHBSupported=1,DoorBellSupported=1,masterUserCodeSync=0,GeofenceStatus=0
DeviceList: OrderedDict([('DeviceInfoBasic', [OrderedDict([('DeviceFlags', 'PromptForInstallerCode=0,AllowUserSlotEditing=0,CalCapable=1,PanelType=8')])])])

**Comment by austinmroczek** [comment]
UserFeatureList --> you're using your master account.  Recommend making a second account in Total Connect that doesn't have account admin permissions to use with Home Assistant.  Not the problem though...

ClientPreferences --> interesting that you have zwave stuff listed there, because I also have a zwave module but nothing listed.  Again, not the problem we're trying to solve...

LocationModuleFlags --> I have AutoSyncEnabled=1 and SyncStatusSupported=1 where you have zeros.  My panel type is 8 and yours is 5...but don't think that's the issue.

My guess is that the AutoSyncEnabled and/or SyncStatusSupported are the reason you're getting the -120 error.

**Comment by codypet** [comment]
Is there something I can change on the panel side or app side to adjust the sync to test to see if I can get it to work?

Thanks for all your help, btw!  Looking forward to 0.107 just to get operating again!

**Comment by austinmroczek** [comment]
I don't know for sure what you can do.  I looked through Total Connect and didn't see any settings that look like an obvious solution, but you could try flipping some settings either on the panel or in Total Connect and see if any of settings above change.

It may also be your panel hardware that is the limiting factor.  Are you on an older model?

It's also possible there is one or more flags that your alarm reporting company controls.  I know for my system, I have to pay extra to get the Smart Scenes enabled for remote access / Total Connect.

**Comment by codypet** [comment]
The panel itself is about 5years old and is the traditional keypad style not the touchscreen.   Im friends with the owner of the security company and the sensors should be communicating.  I know when I log into the app I now get a stream of what sensors open and when after I enables the sensors for 0.106.  When it gets up and running in 0.107 I'll create an aux account for Homeassistant and see if that helps it.

**Comment by austinmroczek** [comment]
Can you give the panel model number, so that way if someone else has the same issue we can figure out if it's a hardware limitation.  It sounds like a Vista model, but hopefully you could find the version number.

**Comment by austinmroczek** [comment]
This fix is in the .107 beta release if you want to try it.  See https://www.home-assistant.io/hassio/installation/ about half way down says how to install beta.

My fix was simple and shouldn't be a problem, but it's possible there would be other issues in the beta release.

**Comment by codypet** [comment]
> Can you give the panel model number, so that way if someone else has the same issue we can figure out if it's a hardware limitation. It sounds like a Vista model, but hopefully you could find the version number.

I don't know where there model number is but it's a Honeywell Lynxplus2 model.

**Comment by codypet** [comment]
Sorry I was out of commission for a little bit as we have a baby with a fever and with what's going on right now, it was kind of a yikes moment.   That said I updated to 0.107 today and we have some interesting results.

I'm back to being operational from the UI however HA states that it cannot determine the status of the panel.   It does disarm the panel when I hit the button, but it won't change the panels state to disarmed in Hass.  Thus, I'm not able to re-arm the panel from Hass.

I do get the sensor's status.  I've attached a screenshot of my lovelace with the entity filter showing the faulted sensors.   
![image](https://user-images.githubusercontent.com/17790157/76995124-c7804480-6925-11ea-99b3-deb138ee4584.png)
The updates you made were pretty successful.  I don't know what the deal is with the panel status, but I'm glad we got most of it working.
